File: clients/windows/src/ChatOS.Core/State/ConversationHistoryStore.cs
using ChatOS.Core.Domain;

namespace ChatOS.Core.State;

public sealed class ConversationHistoryStore
{
    private readonly object _gate = new();
    private readonly Dictionary<string, SessionState> _sessions = new(StringComparer.Ordinal);

    public void MergeCachedTurns(
        IReadOnlyList<ConversationTurn> turns,
        string conversationId)
    {
        lock (_gate)
        {
            var state = StateFor(conversationId);
            Merge(turns, conversationId, state);
        }
    }

    public void MergePage(
        HistoryPage page,
        string conversationId,
        ConversationHistoryPageOrigin origin = ConversationHistoryPageOrigin.Latest)
    {
        lock (_gate)
        {
            var state = StateFor(conversationId);
            var changed = Merge(page.Turns, conversationId, state);
            if (changed &&
                origin == ConversationHistoryPageOrigin.Latest &&
                state.ViewportAnchor?.IsPinnedToBottom == false)
            {
                state.UnreadNewerCount++;
            }

            if (origin == ConversationHistoryPageOrigin.Latest)
            {
                if (page.RequestGeneration >= state.NewestAcceptedLatestGeneration)
                {
                    state.NewestAcceptedLatestGeneration = page.RequestGeneration;
                    if (!state.HasLoadedOlderPage)
                    {
                        state.OlderCursor = page.OlderCursor;
                        state.HasOlder = page.HasOlder;
                    }
                }
            }
            else if (page.RequestGeneration >= state.NewestAcceptedOlderGeneration)
            {
                state.NewestAcceptedOlderGeneration = page.RequestGeneration;
                state.HasLoadedOlderPage = true;
                state.OlderCursor = page.OlderCursor;
                state.HasOlder = page.HasOlder;
            }

            state.SnapshotRevision = Math.Max(state.SnapshotRevision, page.SnapshotRevision);
        }
    }

    public void ApplyRealtime(RealtimeTurnEvent turnEvent, bool userIsReadingOlderContent)
    {
        lock (_gate)
        {
            var conversationId = turnEvent.Turn.ConversationId;
            var state = StateFor(conversationId);
            if (!state.AppliedEventIds.Add(turnEvent.EventId))
            {
                return;
            }

            state.LastAppliedEventSequence = Math.Max(
                state.LastAppliedEventSequence,
                turnEvent.EventSequence);
            var changed = Merge(new[] { turnEvent.Turn }, conversationId, state);
            if (changed && userIsReadingOlderContent)
            {
                state.UnreadNewerCount++;
            }
        }
    }

    public void SetViewportAnchor(ViewportAnchor? anchor, string conversationId)
    {
        lock (_gate)
        {
            var state = StateFor(conversationId);
            state.ViewportAnchor = anchor;
            if (anchor?.IsPinnedToBottom == true)
            {
                state.UnreadNewerCount = 0;
            }
        }
    }

    public void MarkNewerContentRead(string conversationId)
    {
        lock (_gate)
        {
            StateFor(conversationId).UnreadNewerCount = 0;
        }
    }

    public void DiscardOptimisticTurn(string conversationId, string turnId)
    {
        lock (_gate)
        {
            var state = StateFor(conversationId);
            if (state.TurnsById.TryGetValue(turnId, out var turn) && turn.Revision == 0)
            {
                state.TurnsById.Remove(turnId);
            }
        }
    }

    public ConversationHistorySnapshot Snapshot(string conversationId)
    {
        lock (_gate)
        {
            var state = StateFor(conversationId);
            return new ConversationHistorySnapshot(
                conversationId,
                state.TurnsById.Values
                    .OrderBy(static turn => turn, ConversationTurnComparer.Instance)
                    .ToArray(),
                state.OlderCursor,
                state.HasOlder,
                state.SnapshotRevision,
                state.ViewportAnchor,
                state.UnreadNewerCount);
        }
    }

    public void Clear(string conversationId)
    {
        lock (_gate)
        {
            _sessions.Remove(conversationId);
        }
    }

    private SessionState StateFor(string conversationId)
    {
        if (!_sessions.TryGetValue(conversationId, out var state))
        {
            state = new SessionState();
            _sessions[conversationId] = state;
        }

        return state;
    }

    private static bool Merge(
        IReadOnlyList<ConversationTurn> incomingTurns,
        string conversationId,
        SessionState state)
    {
        var changed = false;
        var newlySuperseded = incomingTurns
            .Select(static turn => turn.ProjectExecutionContext?.ReplacedExecutionGroupId?.Trim())
            .Where(static value => !string.IsNullOrEmpty(value))
            .OfType<string>()
            .ToArray();
        if (newlySuperseded.Length > 0)
        {
            state.SupersededExecutionGroupIds.UnionWith(newlySuperseded);
            foreach (var pair in state.TurnsById.ToArray())
            {
                if (state.SupersededExecutionGroupIds.Contains(ExecutionGroupIdentity(pair.Value)) &&
                    pair.Value.IsTaskGraphAvailable)
                {
                    state.TurnsById[pair.Key] = pair.Value with { IsTaskGraphAvailable = false };
                    changed = true;
                }
            }
        }

        foreach (var incoming in incomingTurns)
        {
            if (!string.Equals(incoming.ConversationId, conversationId, StringComparison.Ordinal))
            {
                continue;
            }

            var turn = state.SupersededExecutionGroupIds.Contains(ExecutionGroupIdentity(incoming))
                ? incoming with { IsTaskGraphAvailable = false }
                : incoming;
            if (!state.TurnsById.TryGetValue(turn.Id, out var existing))
            {
                state.TurnsById[turn.Id] = turn;
                changed = true;
            }
            else if (turn.Revision > existing.Revision)
            {
                state.TurnsById[turn.Id] = !existing.IsTaskGraphAvailable
                    ? turn with { IsTaskGraphAvailable = false }
                    : turn;
                changed = true;
            }
        }

        return changed;
    }

    private static string ExecutionGroupIdentity(ConversationTurn turn) =>
        string.IsNullOrWhiteSpace(turn.ProjectExecutionContext?.ExecutionGroupId)
            ? turn.UserMessage.Id
            : turn.ProjectExecutionContext.ExecutionGroupId.Trim();

    private sealed class SessionState
    {
        public Dictionary<string, ConversationTurn> TurnsById { get; } = new(StringComparer.Ordinal);

        public string? OlderCursor { get; set; }

        public bool HasOlder { get; set; }

        public long SnapshotRevision { get; set; }

        public long NewestAcceptedLatestGeneration { get; set; }

        public long NewestAcceptedOlderGeneration { get; set; }

        public bool HasLoadedOlderPage { get; set; }

        public long LastAppliedEventSequence { get; set; }

        public HashSet<string> AppliedEventIds { get; } = new(StringComparer.Ordinal);

        public ViewportAnchor? ViewportAnchor { get; set; }

        public int UnreadNewerCount { get; set; }

        public HashSet<string> SupersededExecutionGroupIds { get; } = new(StringComparer.Ordinal);
    }

    private sealed class ConversationTurnComparer : IComparer<ConversationTurn>
    {
        public static ConversationTurnComparer Instance { get; } = new();

        public int Compare(ConversationTurn? left, ConversationTurn? right)
        {
            if (ReferenceEquals(left, right)) return 0;
            if (left is null) return -1;
            if (right is null) return 1;

            var leftHasDate = left.StartedAt != DateTimeOffset.MinValue;
            var rightHasDate = right.StartedAt != DateTimeOffset.MinValue;
            if (leftHasDate && rightHasDate && left.StartedAt != right.StartedAt)
            {
                return left.StartedAt.CompareTo(right.StartedAt);
            }

            var sequence = left.Sequence.CompareTo(right.Sequence);
            return sequence != 0
                ? sequence
                : string.Compare(left.Id, right.Id, StringComparison.Ordinal);
        }
    }
}

File: clients/macos/Sources/ChatOSCore/ConversationHistoryStore.swift
import Foundation

public actor ConversationHistoryStore {
    private struct SessionState: Sendable {
        var turnsByID: [String: ConversationTurn] = [:]
        var olderCursor: String?
        var hasOlder = false
        var snapshotRevision: Int64 = 0
        var newestAcceptedLatestGeneration: Int64 = 0
        var newestAcceptedOlderGeneration: Int64 = 0
        var hasAcceptedLatestPage = false
        var hasLoadedOlderPage = false
        var lastAppliedEventSequence: Int64 = 0
        var appliedEventIDs: Set<String> = []
        var viewportAnchor: ViewportAnchor?
        var unreadNewerCount = 0
    }

    private var sessions: [String: SessionState] = [:]

    public init() {}

    public func mergeCachedTurns(_ turns: [ConversationTurn], sessionID: String) {
        var state = sessions[sessionID] ?? SessionState()
        merge(
            turns,
            sessionID: sessionID,
            replacingChangedEqualRevisions: false,
            into: &state
        )
        sessions[sessionID] = state
    }

    public func mergePage(
        _ page: HistoryPage,
        sessionID: String,
        origin: ConversationHistoryPageOrigin = .latest
    ) {
        var state = sessions[sessionID] ?? SessionState()
        let acceptsLatestSnapshot = origin == .latest
            && page.requestGeneration >= state.newestAcceptedLatestGeneration
        let mergeResult = merge(
            page.turns,
            sessionID: sessionID,
            replacingChangedEqualRevisions: acceptsLatestSnapshot,
            into: &state
        )

        if state.hasAcceptedLatestPage,
           origin == .latest,
           state.viewportAnchor?.isPinnedToBottom == false,
           mergeResult.newContentCount > 0 {
            state.unreadNewerCount += mergeResult.newContentCount
        }

        switch origin {
        case .latest:
            if page.requestGeneration >= state.newestAcceptedLatestGeneration {
                state.newestAcceptedLatestGeneration = page.requestGeneration
                state.hasAcceptedLatestPage = true
                if !state.hasLoadedOlderPage {
                    state.olderCursor = page.olderCursor
                    state.hasOlder = page.hasOlder
                }
            }
        case .older:
            if page.requestGeneration >= state.newestAcceptedOlderGeneration {
                state.newestAcceptedOlderGeneration = page.requestGeneration
                state.hasLoadedOlderPage = true
                state.olderCursor = page.olderCursor
                state.hasOlder = page.hasOlder
            }
        }
        state.snapshotRevision = max(state.snapshotRevision, page.snapshotRevision)

        sessions[sessionID] = state
    }

    public func applyRealtime(_ event: RealtimeTurnEvent, userIsReadingOlderContent: Bool) {
        let sessionID = event.turn.sessionID
        var state = sessions[sessionID] ?? SessionState()

        guard !state.appliedEventIDs.contains(event.eventID) else {
            return
        }

        state.appliedEventIDs.insert(event.eventID)
        state.lastAppliedEventSequence = max(state.lastAppliedEventSequence, event.eventSequence)
        let mergeResult = merge(
            [event.turn],
            sessionID: sessionID,
            replacingChangedEqualRevisions: false,
            into: &state
        )

        if userIsReadingOlderContent, mergeResult.newContentCount > 0 {
            state.unreadNewerCount += mergeResult.newContentCount
        }

        sessions[sessionID] = state
    }

    public func setViewportAnchor(_ anchor: ViewportAnchor?, sessionID: String) {
        var state = sessions[sessionID] ?? SessionState()
        state.viewportAnchor = anchor
        if anchor?.isPinnedToBottom == true {
            state.unreadNewerCount = 0
        }
        sessions[sessionID] = state
    }

    public func markNewerContentRead(sessionID: String) {
        var state = sessions[sessionID] ?? SessionState()
        state.unreadNewerCount = 0
        sessions[sessionID] = state
    }

    public func discardOptimisticTurn(sessionID: String, turnID: String) {
        var state = sessions[sessionID] ?? SessionState()
        guard state.turnsByID[turnID]?.revision == 0 else { return }
        state.turnsByID.removeValue(forKey: turnID)
        sessions[sessionID] = state
    }

    public func snapshot(sessionID: String) -> ConversationHistorySnapshot {
        let state = sessions[sessionID] ?? SessionState()
        return ConversationHistorySnapshot(
            sessionID: sessionID,
            turns: state.turnsByID.values.sorted(by: ConversationTurn.isOrderedBefore),
            olderCursor: state.olderCursor,
            hasOlder: state.hasOlder,
            snapshotRevision: state.snapshotRevision,
            viewportAnchor: state.viewportAnchor,
            unreadNewerCount: state.unreadNewerCount
        )
    }

    private struct MergeResult {
        var newContentCount = 0
    }

    @discardableResult
    private func merge(
        _ incomingTurns: [ConversationTurn],
        sessionID: String,
        replacingChangedEqualRevisions: Bool,
        into state: inout SessionState
    ) -> MergeResult {
        var result = MergeResult()

        for incomingTurn in incomingTurns {
            var turn = incomingTurn
            guard turn.sessionID == sessionID else { continue }

            guard let existing = state.turnsByID[turn.id] else {
                state.turnsByID[turn.id] = turn
                result.newContentCount += 1
                continue
            }

            if turn.revision > existing.revision
                || (replacingChangedEqualRevisions
                    && turn.revision == existing.revision
                    && turn != existing) {
                if !existing.isTaskGraphAvailable {
                    turn.isTaskGraphAvailable = false
                }
                result.newContentCount += Self.newReplyCount(from: existing, to: turn)
                state.turnsByID[turn.id] = turn
            }
        }

        return result
    }

    private static func newReplyCount(
        from existing: ConversationTurn,
        to incoming: ConversationTurn
    ) -> Int {
        let existingIDs = visibleReplyIDs(for: existing)
        return visibleReplyIDs(for: incoming).subtracting(existingIDs).count
    }

    private static func visibleReplyIDs(for turn: ConversationTurn) -> Set<String> {
        if !turn.assistantReplies.isEmpty {
            return Set(turn.assistantReplies.map(\.id))
        }
        return Set(turn.finalAssistantMessage.map { [$0.id] } ?? [])
    }
}

private extension ConversationTurn {
    static func isOrderedBefore(_ lhs: ConversationTurn, _ rhs: ConversationTurn) -> Bool {
        let lhsHasStartedAt = lhs.startedAt != .distantPast
        let rhsHasStartedAt = rhs.startedAt != .distantPast

        // Some older gateways do not return sequence_no. The mapper can only assign a
        // page-local fallback in that case, so sequence values repeat after loading an
        // older page. A real creation time is therefore the stable cross-page order.
        if lhsHasStartedAt, rhsHasStartedAt, lhs.startedAt != rhs.startedAt {
            return lhs.startedAt < rhs.startedAt
        }

        // Preserve server ordering for legacy records with no usable timestamp and use
        // it as the tie-breaker when multiple turns share the same creation time.
        if lhs.sequence != rhs.sequence {
            return lhs.sequence < rhs.sequence
        }

        return lhs.id < rhs.id
    }
}
